pkg/integrations/v2/eventhandler/eventhandler.go

```diff
@@ -0,0 +1,454 @@
+	ctx, cancel := signal.NotifyContext(ctx, os.Interrupt)
+	defer cancel()
+
+	// todo: figure this out on K8s (PVC, etc.)
+	cacheDir := filepath.Dir(eh.CachePath)
+	// todo: correct perms here? / k8s config
+	if err := os.MkdirAll(cacheDir, 0777); err != nil {
+		level.Error(eh.Log).Log("msg", "Failed to create cache dir", "err", err)
+		cancel()
+	}
+
+	// cache file to store events shipped (prevents double shipping on restart)
+	cacheFile, err := os.OpenFile(eh.CachePath, os.O_RDWR|os.O_CREATE, cacheFileMode)
+	if err != nil {
+		level.Error(eh.Log).Log("msg", "Failed to open or create cache file", "err", err)
+		cancel()
+	}
+	eh.CacheFile = cacheFile
+
+	// attempt to read last timestamp from cache file into a ShippedEvents struct
+	initEvent, err := readInitEvent(eh.CacheFile, eh.Log)
+	if err != nil {
+		level.Error(eh.Log).Log("msg", "Failed to read last event from cache file", "err", err)
+		cancel()
+	}
+	eh.InitEvent = initEvent
+
+	// sync & close cache file
+	defer func() {
+		if err := eh.CacheFile.Close(); err != nil {
+			level.Error(eh.Log).Log("msg", "Failed to close cache file", "err", err)
+			cancel()
+		}
+	}()
+	// todo: do we need this?
+	defer func() {
+		if err := eh.CacheFile.Sync(); err != nil {
+			level.Error(eh.Log).Log("msg", "Failed to sync cache file", "err", err)
+			cancel()
+		}
+	}()
+	// write out .LastEvent struct to cache file on exit
+	defer func() {
+		if err := eh.writeOutLastEvent(); err != nil {
+			level.Error(eh.Log).Log("msg", "Failed to write out last event", "err", err)
+			cancel()
+		}
+	}()
+
+	// todo: is it worth even doing this?
+	go func() {
+		level.Info(eh.Log).Log("msg", "Waiting for cache to sync (initial List of events)")
+		isSynced := cache.WaitForCacheSync(ctx.Done(), eh.EventInformer.HasSynced)
+		if !isSynced {
+			level.Error(eh.Log).Log("msg", "Failed to sync informer cache")
+			// todo: do we want to bail here?
+			return
+		}
+		level.Info(eh.Log).Log("msg", "Informer cache synced")
+	}()
+
+	// start the informer
+	// technically we should prob use the factory here, but since we
+	// only have one informer atm, this likely doesn't matter
+	eh.EventInformer.Run(ctx.Done())
+
+	<-ctx.Done()
+
+	return nil
+}
+
+func readInitEvent(file *os.File, logger log.Logger) (*ShippedEvents, error) {
+	var (
+		initEvent = new(ShippedEvents)
+		// skip first newline
+		cur  int64 = -1
+		char       = make([]byte, 1)
+		line       = ""
+	)
+
+	stat, err := file.Stat()
+	if err != nil {
+		return nil, err
+	}
+	if stat.Size() == 0 {
+		level.Info(logger).Log("msg", "Cache file empty, setting zero-valued initEvent")
+		return initEvent, nil
+	}
+
+	size := stat.Size()
+	for {
+		cur--
+		_, err = file.Seek(cur, 2)
+		if err != nil {
+			return nil, err
+		}
+		_, err = file.Read(char)
+		if err != nil {
+			return nil, err
+		}
+		// newline
+		if char[0] == 10 {
+			break
+		}
+		line = string(char) + line
+		if cur == -size {
+			break
+		}
+	}
```

Comment:
I'm a little thrown off about what's going on here. If you're looking for the most recent JSON object, you actually don't need to do any of this, and you can just call Decode in a loop:

```go
dec := json.NewDecoder(file) 

for {
  err := dec.Decode(&initEvent)
  if errors.Is(err, io.EOF) {
    // EOF is returned after there's no more JSON to read 
  } else if err != nil {
     // Non-EOF error, return 
     return nil, err 
  } 
}

return initEvent, nil 
```

Reply:
(but also I'm skeptical that this file needs to store a series of events instead of just the latest, anyway)

Reply:
Yeah I'm gonna remove this stuff and just flush the latest event struct every interval